File: 백준/Gold/7569. 토마토/토마토.py
from collections import deque
import sys

# H: 높이, N: 행, M: 열
M, N, H = list(map(int, sys.stdin.readline().split()))

graph = []

for h in range(H):
    row = []
    for n in range(N):
        row.append(list(map(int, input().split())))
    graph.append(row)

q = deque()
count = 0

for h in range(H):
    for n in range(N):
        for m in range(M):
            if graph[h][n][m] == 1:
                q.append((h, n, m, 0))

dhdndm = ((1, 0, 0), (-1, 0, 0), (0, 1, 0), (0, -1, 0), (0, 0, 1), (0, 0, -1))

while q:
    cur_h, cur_n, cur_m, cur_count = q.popleft()
    count = max(count, cur_count)
    for h, n, m in dhdndm:
        next_h = cur_h + h
        next_n = cur_n + n
        next_m = cur_m + m
        if next_h >= 0 and next_h < H and next_n >= 0 and next_n < N and next_m >= 0 and next_m < M:
            if graph[next_h][next_n][next_m] == 0:
                graph[next_h][next_n][next_m] = 1
                q.append((next_h, next_n, next_m, cur_count + 1))

def isValid():
    for h in range(H):
        for n in range(N):
            for m in range(M):
                if graph[h][n][m] == 0:
                    return False

    return True

if isValid():
    print(count)
else:
    print(-1)

File: 백준/Gold/7569. 토마토/토마토.swift
import Foundation

let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let M = input[0]
let N = input[1]
let H = input[2]

let dhdrdc = [(0, 0, 1), (0, 0, -1), (0, 1, 0), (0, -1, 0), (1, 0, 0), (-1, 0, 0)]

var visited: [[[Int]]] = .init(
  repeating: .init(
    repeating: .init(
      repeating: -1,
      count: M
    ),
    count: N
  ),
  count: H
)
var board: [[[Int]]] = []
var queue: [(Int, Int, Int)] = []
var front = 0

for height in 0..<H {
  var temp: [[Int]] = []
  for row in 0..<N {
    let input = readLine()!.split(separator: " ").map { Int(String($0))! }
    temp.append(input)

    for col in 0..<M {
      if input[col] == 1 {
        queue.append((height, row, col))
        visited[height][row][col] = 0
      } else if input[col] == -1 {
        visited[height][row][col] = -2
      }
    }
  }

  board.append(temp)
}

while queue.count >= front + 1 {
  let (curH, curR, curC) = queue[front]
  front += 1

  for (dh, dr, dc) in dhdrdc {
    let nextH = curH + dh
    let nextR = curR + dr
    let nextC = curC + dc

    if nextH < 0 || nextH >= H || nextR < 0 || nextR >= N || nextC < 0 || nextC >= M { continue }
    if board[nextH][nextR][nextC] == -1 || visited[nextH][nextR][nextC] != -1 { continue }

    queue.append((nextH, nextR, nextC))
    visited[nextH][nextR][nextC] = visited[curH][curR][curC] + 1
  }
}

var flag = false
var result = 0
out: for height in 0..<H {
  for row in 0..<N {
    for col in 0..<M {
      result = max(result, visited[height][row][col])
      if visited[height][row][col] == -1 {
        flag = true
        break out
      }
    }
  }
}

if flag {
  print(-1)
} else {
  print(result)
}